File: api/_lib/connection-tester.ts
interface TestResult {
  success: boolean;
  error?: string;
  models?: string[];
}

const TIMEOUT = 10000; // 10 seconds

async function fetchWithTimeout(url: string, options: RequestInit = {}): Promise<Response> {
  const controller = new AbortController();
  const id = setTimeout(() => controller.abort(), TIMEOUT);
  try {
    const response = await fetch(url, { ...options, signal: controller.signal });
    return response;
  } finally {
    clearTimeout(id);
  }
}

async function testOpenAI(apiKey: string, baseUrl?: string | null): Promise<TestResult> {
  const url = `${baseUrl || 'https://api.openai.com'}/v1/models`;
  const resp = await fetchWithTimeout(url, {
    headers: { Authorization: `Bearer ${apiKey}` },
  });
  if (!resp.ok) throw new Error(`HTTP ${resp.status}: ${await resp.text()}`);
  const data = await resp.json();
  const modelIds = (data.data || []).map((m: any) => m.id).sort();
  return { success: true, models: modelIds };
}

async function testAnthropic(apiKey: string): Promise<TestResult> {
  const resp = await fetchWithTimeout('https://api.anthropic.com/v1/messages', {
    method: 'POST',
    headers: {
      'x-api-key': apiKey,
      'anthropic-version': '2023-06-01',
      'content-type': 'application/json',
    },
    body: JSON.stringify({
      model: 'claude-sonnet-4-20250514',
      max_tokens: 1,
      messages: [{ role: 'user', content: 'Hi' }],
    }),
  });
  if (!resp.ok) throw new Error(`HTTP ${resp.status}: ${await resp.text()}`);
  return {
    success: true,
    models: [
      'claude-sonnet-4-20250514',
      'claude-opus-4-20250514',
      'claude-haiku-35-20241022',
    ],
  };
}

async function testGoogle(apiKey: string, model?: string | null): Promise<TestResult> {
  const modelName = model || 'gemini-2.0-flash';
  const url = `https://generativelanguage.googleapis.com/v1beta/models/${modelName}:generateContent?key=${apiKey}`;
  const resp = await fetchWithTimeout(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ contents: [{ parts: [{ text: 'Hi' }] }] }),
  });
  if (!resp.ok) throw new Error(`HTTP ${resp.status}: ${await resp.text()}`);
  return {
    success: true,
    models: ['gemini-2.5-pro', 'gemini-2.5-flash', 'gemini-2.0-flash', 'gemini-2.0-flash-lite'],
  };
}

async function testGroq(apiKey: string): Promise<TestResult> {
  const url = 'https://api.groq.com/openai/v1/models';
  const resp = await fetchWithTimeout(url, {
    headers: { Authorization: `Bearer ${apiKey}` },
  });
  if (!resp.ok) throw new Error(`HTTP ${resp.status}: ${await resp.text()}`);
  const data = await resp.json();
  const modelIds = (data.data || []).map((m: any) => m.id).sort();
  return { success: true, models: modelIds };
}

async function testTavily(apiKey: string): Promise<TestResult> {
  const resp = await fetchWithTimeout('https://api.tavily.com/search', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ api_key: apiKey, query: 'test', max_results: 1 }),
  });
  if (!resp.ok) throw new Error(`HTTP ${resp.status}: ${await resp.text()}`);
  return { success: true };
}

async function testSerper(apiKey: string): Promise<TestResult> {
  const resp = await fetchWithTimeout('https://google.serper.dev/search', {
    method: 'POST',
    headers: { 'X-API-KEY': apiKey, 'Content-Type': 'application/json' },
    body: JSON.stringify({ q: 'test' }),
  });
  if (!resp.ok) throw new Error(`HTTP ${resp.status}: ${await resp.text()}`);
  return { success: true };
}

async function testBrave(apiKey: string): Promise<TestResult> {
  const resp = await fetchWithTimeout(
    'https://api.search.brave.com/res/v1/web/search?q=test',
    {
      headers: { 'X-Subscription-Token': apiKey, Accept: 'application/json' },
    }
  );
  if (!resp.ok) throw new Error(`HTTP ${resp.status}: ${await resp.text()}`);
  return { success: true };
}

type TesterFn = (apiKey: string, baseUrlOrModel?: string | null) => Promise<TestResult>;

const LLM_TESTERS: Record<string, TesterFn> = {
  openai: (apiKey, baseUrl) => testOpenAI(apiKey, baseUrl),
  anthropic: (apiKey) => testAnthropic(apiKey),
  google: (apiKey, model) => testGoogle(apiKey, model),
  gemini: (apiKey, model) => testGoogle(apiKey, model),
  groq: (apiKey) => testGroq(apiKey),
};

const SEARCH_TESTERS: Record<string, TesterFn> = {
  tavily: (apiKey) => testTavily(apiKey),
  serper: (apiKey) => testSerper(apiKey),
  brave: (apiKey) => testBrave(apiKey),
};

export async function testLLMConnection(
  name: string,
  apiKey: string,
  baseUrl?: string | null,
  model?: string | null
): Promise<TestResult> {
  const tester = LLM_TESTERS[name.toLowerCase()];
  if (!tester) return { success: false, error: `Unknown LLM provider: ${name}` };
  try {
    return await tester(apiKey, baseUrl || model);
  } catch (e: any) {
    if (e.name === 'AbortError') return { success: false, error: `Connection timed out after ${TIMEOUT / 1000}s` };
    return { success: false, error: e.message };
  }
}

export async function testSearchConnection(
  name: string,
  apiKey: string
): Promise<TestResult> {
  const tester = SEARCH_TESTERS[name.toLowerCase()];
  if (!tester) return { success: false, error: `Unknown search provider: ${name}` };
  try {
    return await tester(apiKey);
  } catch (e: any) {
    if (e.name === 'AbortError') return { success: false, error: `Connection timed out after ${TIMEOUT / 1000}s` };
    return { success: false, error: e.message };
  }
}

File: backend/config_service/services/connection_tester.py
"""Connection testing logic for LLM and search providers."""

import httpx

from ..models.schemas import TestConnectionRequest, TestConnectionResponse

TIMEOUT = 10.0  # seconds


# ---------------------------------------------------------------------------
# Individual provider testers
# ---------------------------------------------------------------------------

async def _test_openai(api_key: str, base_url: str | None = None, **_kwargs) -> TestConnectionResponse:
    """Test OpenAI connection by listing available models."""
    url = f"{base_url}/v1/models" if base_url else "https://api.openai.com/v1/models"
    async with httpx.AsyncClient(timeout=TIMEOUT) as client:
        resp = await client.get(url, headers={"Authorization": f"Bearer {api_key}"})
        resp.raise_for_status()
        data = resp.json()
        model_ids = [m["id"] for m in data.get("data", [])]
        return TestConnectionResponse(success=True, models=sorted(model_ids))


async def _test_anthropic(api_key: str, **_kwargs) -> TestConnectionResponse:
    """Test Anthropic connection by sending a tiny messages request."""
    url = "https://api.anthropic.com/v1/messages"
    headers = {
        "x-api-key": api_key,
        "anthropic-version": "2023-06-01",
        "content-type": "application/json",
    }
    body = {
        "model": "claude-sonnet-4-20250514",
        "max_tokens": 1,
        "messages": [{"role": "user", "content": "Hi"}],
    }
    async with httpx.AsyncClient(timeout=TIMEOUT) as client:
        resp = await client.post(url, json=body, headers=headers)
        resp.raise_for_status()
        return TestConnectionResponse(
            success=True,
            models=[
                "claude-sonnet-4-20250514",
                "claude-opus-4-20250514",
                "claude-haiku-35-20241022",
            ],
        )


async def _test_google(api_key: str, model: str | None = None, **_kwargs) -> TestConnectionResponse:
    """Test Google Gemini connection by sending a tiny generateContent request."""
    model_name = model or "gemini-2.0-flash"
    url = (
        f"https://generativelanguage.googleapis.com/v1beta/models/{model_name}"
        f":generateContent?key={api_key}"
    )
    body = {"contents": [{"parts": [{"text": "Hi"}]}]}
    async with httpx.AsyncClient(timeout=TIMEOUT) as client:
        resp = await client.post(url, json=body)
        resp.raise_for_status()
        return TestConnectionResponse(
            success=True,
            models=[
                "gemini-2.5-pro",
                "gemini-2.5-flash",
                "gemini-2.0-flash",
                "gemini-2.0-flash-lite",
            ],
        )


async def _test_groq(api_key: str, **_kwargs) -> TestConnectionResponse:
    """Test Groq connection by listing available models."""
    url = "https://api.groq.com/openai/v1/models"
    async with httpx.AsyncClient(timeout=TIMEOUT) as client:
        resp = await client.get(url, headers={"Authorization": f"Bearer {api_key}"})
        resp.raise_for_status()
        data = resp.json()
        model_ids = [m["id"] for m in data.get("data", [])]
        return TestConnectionResponse(success=True, models=sorted(model_ids))


async def _test_ollama(base_url: str | None = None, **_kwargs) -> TestConnectionResponse:
    """Test Ollama connection by listing local models."""
    url = f"{base_url or 'http://localhost:11434'}/api/tags"
    async with httpx.AsyncClient(timeout=TIMEOUT) as client:
        resp = await client.get(url)
        resp.raise_for_status()
        data = resp.json()
        model_names = [m["name"] for m in data.get("models", [])]
        return TestConnectionResponse(success=True, models=sorted(model_names))


async def _test_tavily(api_key: str, **_kwargs) -> TestConnectionResponse:
    """Test Tavily search connection."""
    url = "https://api.tavily.com/search"
    body = {"api_key": api_key, "query": "test", "max_results": 1}
    async with httpx.AsyncClient(timeout=TIMEOUT) as client:
        resp = await client.post(url, json=body)
        resp.raise_for_status()
        return TestConnectionResponse(success=True)


async def _test_serper(api_key: str, **_kwargs) -> TestConnectionResponse:
    """Test Serper search connection."""
    url = "https://google.serper.dev/search"
    headers = {"X-API-KEY": api_key, "Content-Type": "application/json"}
    body = {"q": "test"}
    async with httpx.AsyncClient(timeout=TIMEOUT) as client:
        resp = await client.post(url, json=body, headers=headers)
        resp.raise_for_status()
        return TestConnectionResponse(success=True)


async def _test_brave(api_key: str, **_kwargs) -> TestConnectionResponse:
    """Test Brave Search connection."""
    url = "https://api.search.brave.com/res/v1/web/search"
    headers = {"X-Subscription-Token": api_key, "Accept": "application/json"}
    async with httpx.AsyncClient(timeout=TIMEOUT) as client:
        resp = await client.get(url, params={"q": "test"}, headers=headers)
        resp.raise_for_status()
        return TestConnectionResponse(success=True)


# ---------------------------------------------------------------------------
# Registry & main entry point
# ---------------------------------------------------------------------------

_TESTERS = {
    "openai": _test_openai,
    "anthropic": _test_anthropic,
    "google": _test_google,
    "gemini": _test_google,
    "groq": _test_groq,
    "ollama": _test_ollama,
    "tavily": _test_tavily,
    "serper": _test_serper,
    "brave": _test_brave,
}


async def test_connection(request: TestConnectionRequest) -> TestConnectionResponse:
    """Test a provider connection without persisting anything.

    Routes to the appropriate provider-specific tester based on ``request.name``.
    """
    provider_name = request.name.lower()
    tester = _TESTERS.get(provider_name)
    if tester is None:
        return TestConnectionResponse(
            success=False,
            error=f"Unknown provider: {request.name}. Supported: {', '.join(_TESTERS.keys())}",
        )

    try:
        return await tester(
            api_key=request.api_key,
            base_url=request.base_url,
            model=request.model,
        )
    except httpx.HTTPStatusError as exc:
        return TestConnectionResponse(
            success=False,
            error=f"HTTP {exc.response.status_code}: {exc.response.text[:300]}",
        )
    except httpx.ConnectError:
        return TestConnectionResponse(
            success=False,
            error=f"Connection refused – is {request.name} reachable?",
        )
    except httpx.TimeoutException:
        return TestConnectionResponse(
            success=False,
            error=f"Connection to {request.name} timed out after {TIMEOUT}s",
        )
    except Exception as exc:  # noqa: BLE001
        return TestConnectionResponse(success=False, error=str(exc))
